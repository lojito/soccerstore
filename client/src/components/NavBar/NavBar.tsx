import { type FC } from "react";
import { MdOutlineShoppingCart } from "react-icons/md";
import { NavLink as BaseNavLink } from "react-router-dom";
import { styled } from "styled-components";
import Filter, { Category } from "../Filter/Filter";
import Logo from "../Logo/Logo";
import Search from "../Search/Search";

const Nav = styled.header`
  height: 140px;
  background-color: black;
  position: fixed;
  top: 0px;
  max-width: 1200px;
  width: 100%;
  padding-top: 10px;
  z-index: 100;

  @media screen and (max-width: 720px) {
    height: 180px;
  }

  @media screen and (max-width: 410px) {
    height: 210px;
  }

  @media screen and (max-width: 310px) {
    height: 280px;
  }

  @media screen and (max-width: 280px) {
    height: 280px;
  }
`;

const Ul = styled.ul`
  list-style: none;
  padding-left: 0;
  margin: 0;
  display: inline-block;
  text-align: right;
  width: 30%;
  vertical-align: top;
  padding-top: 12px;

  @media screen and (max-width: 950px) {
    width: 40%;
  }

  @media screen and (max-width: 720px) {
    width: 100%;
    text-align: center;
  }
`;

const Li = styled.li`
  display: inline-block;
  margin-right: 10px;
  cursor: pointer;

  & svg {
    color: #fff;
    width: 27px;
    height: 27px;
    vertical-align: middle;
  }
`;

const Button = styled.button`
  cursor: pointer;
  border: 1px solid black;
  border-radius: 6px;
  padding: 5px;
`;

const NavLink = styled(BaseNavLink)`
  color: #fff;
  text-decoration: none;
  &.active {
    font-weight: 1000;
  }
`;

const Welcome = styled.span`
  color: #fff;
  cursor: auto;
`;

type Props = {
  name: string;
  isLoggedIn: boolean;
  onLogout: () => void;
  onSearch: (searchTerm: string) => void;
  onFilterCategory: (filterCategory: Category) => void;
};

const NavBar: FC<Props> = ({
  name,
  isLoggedIn,
  onLogout,
  onSearch,
  onFilterCategory,
}) => {
  return (
    <Nav>
      <Logo />
      <Search onSearch={(searchTerm: string) => onSearch(searchTerm)} />
      <Ul>
        {isLoggedIn ? (
          <>
            <Li key="name">
              <Welcome>Bienvenido, {name.substring(0, 5) + "..."}</Welcome>
            </Li>
            <Li key="logout">
              <Button onClick={onLogout}>Cerrar sesión</Button>
            </Li>
            <Li key="cart">
              <NavLink to="/cart">
                <MdOutlineShoppingCart title="Lista de productos a comprar." />
              </NavLink>
            </Li>
          </>
        ) : (
          <>
            <Li key="login">
              <NavLink to="/login">Iniciar sesión</NavLink>
            </Li>
            <Li key="signup">
              <NavLink to="/signup">Regístrate</NavLink>
            </Li>
          </>
        )}
      </Ul>
      <Filter
        isLoggedIn={isLoggedIn}
        onFilterCategory={(filterCategory: Category) =>
          onFilterCategory(filterCategory)
        }
      />
    </Nav>
  );
};

export default NavBar;
