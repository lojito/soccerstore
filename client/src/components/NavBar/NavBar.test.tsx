import { render, screen } from "@testing-library/react";
import NavBar from "./NavBar";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";

vi.mock("../Logo/Logo");

vi.mock("../Search/Search", () => ({
  default: ({ onSearch }: { onSearch: (searchTerm: string) => void }) => {
    return (
      <input
        type="text"
        id="text"
        name="search"
        placeholder="Buscar productos"
        onChange={(e) => onSearch(e.target.value)}
      />
    );
  },
}));

vi.mock("../Filter/Filter", () => ({
  default: ({
    onFilterCategory,
  }: {
    onFilterCategory: (filterCategory: string) => void;
  }) => {
    return (
      <button aria-label="todos" onClick={() => onFilterCategory("todos")}>
        Todos
      </button>
    );
  },
}));

describe("NavBar", () => {
  interface Props {
    name?: string;
    isLoggedIn?: boolean;
  }

  const renderNavBar = ({ name = "", isLoggedIn = false }: Props = {}) => {
    const onSearch = vi.fn();
    const onLogout = vi.fn();
    const onFilterCategory = vi.fn();

    const utils = render(
      <MemoryRouter>
        <NavBar
          name={name}
          isLoggedIn={isLoggedIn}
          onLogout={onLogout}
          onSearch={onSearch}
          onFilterCategory={onFilterCategory}
        />
      </MemoryRouter>
    );

    const search = screen.getByRole("textbox");
    const filterButton = screen.getByRole("button", { name: "todos" });
    const loginLink = screen.queryByRole("link", { name: "Iniciar sesión" });
    const signUpLink = screen.queryByRole("link", { name: "Regístrate" });
    const logOutButton = screen.queryByRole("button", {
      name: "Cerrar sesión",
    });

    return {
      ...utils,
      search,
      filterButton,
      loginLink,
      signUpLink,
      logOutButton,
      onSearch,
      onLogout,
      onFilterCategory,
    };
  };

  it("should render correctly", () => {
    const { asFragment } = renderNavBar();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should call the onSearch prop when entering a search term", async () => {
    const user = userEvent.setup();
    const { onSearch } = renderNavBar();

    const search = screen.getByRole("textbox");
    await user.type(search, "Rojo");

    expect(onSearch).toHaveBeenCalled();
  });

  it("should call the onFilterCategory prop when clicing on a filter button", async () => {
    const user = userEvent.setup();
    const { onFilterCategory, filterButton } = renderNavBar();

    await user.click(filterButton);

    expect(onFilterCategory).toHaveBeenCalled();
  });

  it("should show the 'Iniciar sesión' and Regístrate links when the user is not logged in", async () => {
    const { loginLink, signUpLink } = renderNavBar();

    expect(loginLink).toBeInTheDocument();
    expect(signUpLink).toBeInTheDocument();
  });

  it("should not show the 'Cerrar sesión' button when the user is not logged in", async () => {
    const { logOutButton } = renderNavBar();

    expect(logOutButton).not.toBeInTheDocument();
  });

  it("should not show the 'Iniciar sesión' and Regístrate links when the user is logged in", async () => {
    const { loginLink, signUpLink } = renderNavBar({ isLoggedIn: true });

    expect(loginLink).not.toBeInTheDocument();
    expect(signUpLink).not.toBeInTheDocument();
  });

  it("should show the 'Cerrar sesión' button when the user is logged in", async () => {
    const { logOutButton } = renderNavBar({ isLoggedIn: true });

    expect(logOutButton).toBeInTheDocument();
  });
});
