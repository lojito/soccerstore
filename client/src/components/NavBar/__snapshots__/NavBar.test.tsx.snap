// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavBar > should render correctly 1`] = `
<DocumentFragment>
  <header
    class="sc-guDLey iuYCVI"
  >
    <input
      id="text"
      name="search"
      placeholder="Buscar productos"
      type="text"
    />
    <ul
      class="sc-dmyCSP bsLMCO"
    >
      <li
        class="sc-hLQSwg gdwXMY"
      >
        <a
          class="sc-jTQCzO dkUpBw"
          href="/login"
        >
          Iniciar sesión
        </a>
      </li>
      <li
        class="sc-hLQSwg gdwXMY"
      >
        <a
          class="sc-jTQCzO dkUpBw"
          href="/signup"
        >
          Regístrate
        </a>
      </li>
    </ul>
    <button
      aria-label="todos"
    >
      Todos
    </button>
  </header>
</DocumentFragment>
`;
